### DISCUSSION TITLE
unexpected outcome of `cooperative_tradeoff` with two-strain community of default E.coli models

### DISCUSSION THREAD

**Discussion by justuswfink** [micom-dev/micom#158] [Q&A]
Hi Christian, dear MICOM community,

I have a case where the 'cooperative_tradeoff' method does not work as expected and wanted to share it here, to see if anyone can reproduce it. Basically, the solution returned by `cooperative_tradeoff` only allows one E.coli two grow, when I put two default E.coli models together. Here is the code:

```
from micom import Community

# let's create a simple community with two strains
files = [['Ecoli_old', './iJO1366.xml'],
        ['Ecoli_new', './iML1515.xml']]
taxa = pd.DataFrame(files, columns = ['id', 'file'])
com_mixed = Community(taxa, solver = 'gurobi') 

# let's compute the solution
sol_gurobi = com_mixed.cooperative_tradeoff(fluxes = True, pfba = True)
print('The solver has exited with the status: ' +  sol_gurobi.status)
print(sol_gurobi.members)
```

Here is the output: 
<img width="785" alt="Screenshot 2024-01-29 at 09 29 50" src="https://github.com/micom-dev/micom/assets/144257498/8f1a0b31-305b-414d-8cb7-bc5fc757f76c">

This outcome is confusing for me, because both E. coli models should be able to grow. I checked for their individual growth using the built-in function `com_mixed.optimize_all()` and can see the maximum growth rate for each strain: 

<img width="483" alt="Screenshot 2024-01-29 at 09 30 50" src="https://github.com/micom-dev/micom/assets/144257498/8bba0738-9dfe-4b6b-ac56-2b1230e55088">

I would expect the `cooperative tradeoff` method to find the balance here in the middle, but it seems like it allocates all to the Ecoli strain with the higher maximum rate. Am I missing something? 

I have checked a couple of alternatives here:
- I get the same asymmetric solution when I use OSQP as the solver (return status infeasible) and CPLEX as the solver (return status aborted). This suggests that something is ill-defined about this optimization problem more generally, and GUROBI just forgot to complain (and kept a nice exit status).
- The outcome does not depend on the setting of pFBA=True/False or fluxes = True/False. 
- Interestingly, the error only appears when the two E.coli models are different. When I create a community from fusing two identical versions, I do not get this error. 

For example, replaying  the code above with 

```
files = [['Strain_1', './iJO1366.xml'],
         ['Strain_2', './iJO1366.xml']]
taxa = pd.DataFrame(files, columns = ['id', 'file'])
```
gives the nicely balanced solution that I would expect.

<img width="535" alt="Screenshot 2024-01-29 at 09 35 49" src="https://github.com/micom-dev/micom/assets/144257498/ed8a9ec6-764b-4b90-a7b7-ed1620a165e6">

Happy to discuss it further and provide more details! This observation first came up when we were creating genetically modified models of E. coli and looking at them in communities. MICOM seems like such a great way to test there growth and metabolism, and I am glad to put in some extra work to make sure this scenario works. All the best,

Justus

**Comment by cdiener** [comment]
Since one taxon is strictly dominant, you would have to use a tradeoff <1.0 (`com.cooperative_tradeoff(fraction=...)`. If you don't specify one it still requires the maximum theoretical community growth rate to be reached (`fraction=1.0`) which is only the case if `E_coli_old` grows at its maximum rate. If you set a smaller tradeoff you will see both taxa grow. There is some more discussion at https://journals.asm.org/doi/10.1128/msystems.01270-22 .

**Reply by justuswfink** [reply]
Thanks for pointing this out, I didn't consider the `fraction=` parameter and now see how it should work! This has fixed the issue for the solvers 'OSQP' and 'CPLEX'.

I am still struggling with the same outcome in 'GUROBI', though. When I go to a lower fraction, only one of the strains grows. 

<img width="796" alt="Screenshot 2024-01-29 at 19 39 03" src="https://github.com/micom-dev/micom/assets/144257498/81f72107-dc79-48a0-8b70-ab1a5d34e3ee">

I have no clue why GUROBI would be worse than OSQP here. Going forward, I will try to use the CPLEX solver then!

**Reply by cdiener** [reply]
That looks like a bug for sure. Maybe some constraints are not getting reset. I will try to reproduce this with the models you used.
